### ISSUE TITLE
Concurrent map access in daemonset checker

### ISSUE THREAD

**Issue by integrii** [Comcast/kuberhealthy#181]
Very rarely, the daemonset checker crashes Kuberhealthy due to a concurrent map access.  Here is the stack trace:

```
fatal error: concurrent map read and map write

goroutine 12 [running]:
runtime.throw(0x1226f97, 0x21)
	/usr/local/go/src/runtime/panic.go:617 +0x72 fp=0xc001275310 sp=0xc0012752e0 pc=0x42d7c2
runtime.mapaccess2(0x10bf780, 0xc0001c8930, 0xc0012753d0, 0xc00020e158, 0x115cd01)
	/usr/local/go/src/runtime/map.go:472 +0x272 fp=0xc001275358 sp=0xc001275310 pc=0x40f1e2
k8s.io/apimachinery/pkg/runtime.(*Scheme).ObjectKinds(0xc000158a10, 0x13af1c0, 0xc00085da00, 0x0, 0x121772d, 0x10, 0xc00098d880, 0x13c7d20, 0xc000a152c0)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20181127025237-2b1284ed4c93/pkg/runtime/scheme.go:260 +0x252 fp=0xc001275420 sp=0xc001275358 pc=0x7f4462
k8s.io/apimachinery/pkg/runtime.(*parameterCodec).EncodeParameters(0xc0000cee40, 0x13af1c0, 0xc00085da00, 0x12184ff, 0x11, 0x120f90e, 0x2, 0x26, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20181127025237-2b1284ed4c93/pkg/runtime/codec.go:178 +0x62 fp=0xc001275518 sp=0xc001275420 pc=0x7e1642
k8s.io/client-go/rest.(*Request).SpecificallyVersionedParams(0xc001426000, 0x13af1c0, 0xc00085da00, 0x13af6c0, 0xc0000cee40, 0x12184ff, 0x11, 0x120f90e, 0x2, 0x10)
	/go/pkg/mod/k8s.io/client-go@v10.0.0+incompatible/rest/request.go:327 +0xad fp=0xc001275620 sp=0xc001275518 pc=0xe514bd
k8s.io/client-go/rest.(*Request).VersionedParams(...)
	/go/pkg/mod/k8s.io/client-go@v10.0.0+incompatible/rest/request.go:320
github.com/Comcast/kuberhealthy/pkg/khstatecrd.(*KuberhealthyStateClient).Get(0xc00098db00, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1211a1a, 0x8, ...)
	/go/src/github.com/Comcast/kuberhealthy/pkg/khstatecrd/functions.go:70 +0x196 fp=0xc001275700 sp=0xc001275620 pc=0xe93106
main.ensureCRDExists(0xc000b0c3c0, 0x26, 0xc00098db00, 0x2, 0xc0000c9300)
	/go/src/github.com/Comcast/kuberhealthy/cmd/kuberhealthy/crd.go:77 +0x15e fp=0xc001275c00 sp=0xc001275700 pc=0xfe021e
main.(*Kuberhealthy).storeCheckState(0xc000236320, 0xc000b0c3c0, 0x26, 0x1, 0x1d72c20, 0x0, 0x0, 0x1213b5d, 0xb, 0x0, ...)
	/go/src/github.com/Comcast/kuberhealthy/cmd/kuberhealthy/kuberhealthy.go:275 +0xb8 fp=0xc001275cc8 sp=0xc001275c00 pc=0xfe2ab8
main.(*Kuberhealthy).runCheck(0xc000236320, 0xc000328230, 0x13de100, 0xc00024e700)
	/go/src/github.com/Comcast/kuberhealthy/cmd/kuberhealthy/kuberhealthy.go:257 +0x7b6 fp=0xc001275fc0 sp=0xc001275cc8 pc=0xfe26e6
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1337 +0x1 fp=0xc001275fc8 sp=0xc001275fc0 pc=0x459cb1
created by main.(*Kuberhealthy).StartChecks
	/go/src/github.com/Comcast/kuberhealthy/cmd/kuberhealthy/kuberhealthy.go:124 +0xf9

```
 - Kubernetes Version: 1.9.7
 - Kuberhealthy: 1.0.2

**Comment by shuttr** [comment]
@integrii Just when you thought I was gone... :)  This is EKS 1.12.6... no more openshift!

```
ime="2019-06-28T16:32:18Z" level=info msg="Setting state of check DnsStatusChecker to true []"
time="2019-06-28T16:32:18Z" level=info msg="Setting state of check PodRestartChecker namespace grafana to true []"
time="2019-06-28T16:32:18Z" level=info msg="Setting state of check PodRestartChecker namespace kuberhealthy to true []"
time="2019-06-28T16:32:18Z" level=info msg="Setting state of check PodRestartChecker namespace default to true []"
time="2019-06-28T16:32:18Z" level=info msg="Setting state of check PodRestartChecker namespace prometheus to true []"
time="2019-06-28T16:32:18Z" level=info msg="Setting state of check PodRestartChecker namespace kube-system to true []"
fatal error: concurrent map read and map write

goroutine 15 [running]:
runtime.throw(0x12431b3, 0x21)
	/usr/local/go/src/runtime/panic.go:617 +0x72 fp=0xc00054ada0 sp=0xc00054ad70 pc=0x42c002
runtime.mapaccess2(0x10d8ae0, 0xc000172960, 0xc00054ae60, 0xc0001b8d40, 0x1190d01)
	/usr/local/go/src/runtime/map.go:472 +0x284 fp=0xc00054ade8 sp=0xc00054ada0 pc=0x40d6a4
k8s.io/apimachinery/pkg/runtime.(*Scheme).ObjectKinds(0xc000110a80, 0x13caf80, 0xc0000cc6e0, 0x1, 0xc00054af18, 0x40d491, 0xc00054afa0, 0x3e86bbd6, 0x4a72aa4275040280)
```

**Comment by shuttr** [comment]
So basically, I see two threads referring to the khstates CRD...   Not super seasoned with Go, but I'll have to read the code and understand the routines to gather if there is any implied mutex here or if one should be implemented in code using IPC.

```
$ cloudwatch_log_grabber.sh  "shuttr.*containers" "kuberhealthy-bf546dc4c-7t66p_kuberhealthy_kuberhealthy-f289.*" | sed 1,2d   | jq -r '.events[].message | fromjson | .log' | sed '/^$/d' | tail -275
<SNIP>
time="2019-06-30T13:49:20Z" level=info msg="Setting state of check DnsStatusChecker to true []"
time="2019-06-30T13:49:35Z" level=info msg="Running check: DnsStatusChecker"
time="2019-06-30T13:49:35Z" level=info msg="Running check: ComponentStatusChecker"
time="2019-06-30T13:49:35Z" level=info msg="Running DNS checker"
time="2019-06-30T13:49:35Z" level=info msg="Running check: PodStatusChecker namespace kube-system"
time="2019-06-30T13:49:35Z" level=info msg="DNS Checker testing kubernetes.default"
time="2019-06-30T13:49:35Z" level=info msg="DNS Checker determined that kubernetes.default was OK."
time="2019-06-30T13:49:35Z" level=info msg="Setting state of check DnsStatusChecker to true []"
time="2019-06-30T13:49:35Z" level=info msg="Setting state of check ComponentStatusChecker to true []"
fatal error: concurrent map read and map write
goroutine 34 [running]:
runtime.throw(0x12431b3, 0x21)
	/usr/local/go/src/runtime/panic.go:617 +0x72 fp=0xc000134e98 sp=0xc000134e68 pc=0x42c002
runtime.mapaccess2(0x10d8ae0, 0xc000168960, 0xc000134f58, 0xc0001ac1a0, 0x1190d01)
	/usr/local/go/src/runtime/map.go:472 +0x284 fp=0xc000134ee0 sp=0xc000134e98 pc=0x40d6a4
k8s.io/apimachinery/pkg/runtime.(*Scheme).ObjectKinds(0xc000106a80, 0x13caf80, 0xc000304000, 0x1, 0xc000135010, 0x40d491, 0xc000135098, 0x3887f7a8, 0x201b2b991c6cb6df)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190221213512-86fb29eff628/pkg/runtime/scheme.go:260 +0x252 fp=0xc000134fa8 sp=0xc000134ee0 pc=0x7f2dc2
k8s.io/apimachinery/pkg/runtime.UseOrCreateObject(0x13cf480, 0xc000106a80, 0x13c5260, 0xc000106a80, 0xc000040380, 0x11, 0xc000040392, 0x2, 0xc0000403a0, 0x11, ...)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190221213512-86fb29eff628/pkg/runtime/codec.go:83 +0xf4 fp=0xc0001350e8 sp=0xc000134fa8 pc=0x7df314
k8s.io/apimachinery/pkg/runtime/serializer/json.(*Serializer).Decode(0xc000076dc0, 0xc000024000, 0x1ea, 0x600, 0x0, 0x13caf80, 0xc000304000, 0xc0004528c0, 0x62, 0x1, ...)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190221213512-86fb29eff628/pkg/runtime/serializer/json/json.go:211 +0x74c fp=0xc0001353d8 sp=0xc0001350e8 pc=0x8952fc
k8s.io/apimachinery/pkg/runtime/serializer/versioning.DirectDecoder.Decode(0x13c5300, 0xc000076dc0, 0xc000024000, 0x1ea, 0x600, 0x0, 0x13caf80, 0xc000304000, 0x0, 0xc0004b1290, ...)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190221213512-86fb29eff628/pkg/runtime/serializer/versioning/versioning.go:275 +0x97 fp=0xc000135470 sp=0xc0001353d8 pc=0x9444f7
k8s.io/apimachinery/pkg/runtime/serializer/versioning.(*DirectDecoder).Decode(0xc0003755d0, 0xc000024000, 0x1ea, 0x600, 0x0, 0x13caf80, 0xc000304000, 0x0, 0x0, 0xe5a2d0, ...)
	<autogenerated>:1 +0x9f fp=0xc0001354e8 sp=0xc000135470 pc=0x9446ef
k8s.io/client-go/rest.Result.Into(0xc000024000, 0x1ea, 0x600, 0xc000584c70, 0x10, 0x0, 0x0, 0xc8, 0x13c6520, 0xc0003755d0, ...)
	/go/pkg/mod/k8s.io/client-go@v10.0.0+incompatible/rest/request.go:1106 +0x165 fp=0xc0001355a8 sp=0xc0001354e8 pc=0xe56b15
github.com/Comcast/kuberhealthy/pkg/khstatecrd.(*KuberhealthyStateClient).Get(0xc00037f000, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x122d9d2, 0x8, ...)
<SNIP>
goroutine 30 [runnable]:
reflect.(*rtype).In(0x10ea6c0, 0x0, 0x0, 0x0)
	/usr/local/go/src/reflect/type.go:956 +0xd6
k8s.io/apimachinery/pkg/conversion.ConversionFuncs.Add(0xc000168a50, 0xc000168a80, 0xc000251338, 0x1, 0x1, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190221213512-86fb29eff628/pkg/conversion/converter.go:176 +0x14c
k8s.io/apimachinery/pkg/conversion.(*Converter).RegisterConversionFunc(...)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190221213512-86fb29eff628/pkg/conversion/converter.go:376
k8s.io/apimachinery/pkg/runtime.(*Scheme).AddConversionFuncs(0xc000106a80, 0xc000251388, 0x1d, 0x1d, 0xd610104e4ccb9185, 0x403a8f)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190221213512-86fb29eff628/pkg/runtime/scheme.go:343 +0xa5
k8s.io/apimachinery/pkg/apis/meta/v1.AddConversionFuncs(0xc000106a80, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190221213512-86fb29eff628/pkg/apis/meta/v1/conversion.go:33 +0x3c6
k8s.io/apimachinery/pkg/apis/meta/v1.AddToGroupVersion(0xc000106a80, 0x1234593, 0x11, 0x122b875, 0x2)
	/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20190221213512-86fb29eff628/pkg/apis/meta/v1/register.go:75 +0x6b4
github.com/Comcast/kuberhealthy/pkg/khstatecrd.addKnownTypes(0xc000106a80, 0x0, 0x0)
```

**Comment by shuttr** [comment]
More detail from the API Audit Logs...

This is from one of the two API servers EKS provides. 

It seems that, given that the sourceIP is the same across the two API servers, the same pod restarted and quickly started working again.  The panic was at 13:49:35Z and the deluge of updates on `khstate` happened on auditlog `7db` immediately afterwards.  The last successful write from the previous running of that pod was at 13:49:20Z as seen in `5d6`.

```json
$ cloudwatch_log_grabber.sh  "shuttr.*cluster" "kube-apiserver-audit-5d6.*" "2019-06-30T13:49:20Z" "2019-06-30T13:49:45Z" | sed 1,2d | jq -r '.events[].message | fromjson | select(.objectRef.apiGroup == "comcast.github.io") | select(.verb == "update")' | jq -c 
{"kind":"Event","apiVersion":"audit.k8s.io/v1beta1","metadata":{"creationTimestamp":"2019-06-30T13:49:20Z"},"level":"Metadata","timestamp":"2019-06-30T13:49:20Z","auditID":"7e897c5f-0b7d-4146-95b4-42b49f96df48","stage":"ResponseComplete","requestURI":"/apis/comcast.github.io/v1/namespaces/kuberhealthy/khstates/dnsstatuschecker","verb":"update","user":{"username":"system:serviceaccount:kuberhealthy:kuberhealthy","uid":"9af9ab58-99eb-11e9-b05b-02ca7bb2d18c","groups":["system:serviceaccounts","system:serviceaccounts:kuberhealthy","system:authenticated"]},"sourceIPs":["10.17.21.156"],"userAgent":"kuberhealthy/v0.0.0 (linux/amd64) kubernetes/$Format","objectRef":{"resource":"khstates","namespace":"kuberhealthy","name":"dnsstatuschecker","apiGroup":"comcast.github.io","apiVersion":"v1"},"responseStatus":{"metadata":{},"code":200},"requestReceivedTimestamp":"2019-06-30T13:49:20.609021Z","stageTimestamp":"2019-06-30T13:49:20.612670Z","annotations":{"authorization.k8s.io/decision":"allow","authorization.k8s.io/reason":"RBAC: allowed by RoleBinding \"kuberhealthy/kuberhealthy\" of Role \"kuberhealthy\" to ServiceAccount \"kuberhealthy/kuberhealthy\""}}
```

And this is from the other:
```json
$ cloudwatch_log_grabber.sh  "shuttr.*cluster" "kube-apiserver-audit-7db.*" "2019-06-30T13:49:20Z" "2019-06-30T13:49:45Z" | sed 1,2d | jq -r '.events[].message | fromjson | select(.objectRef.apiGroup == "comcast.github.io") | select(.verb == "update")' | jq -c 
{"kind":"Event","apiVersion":"audit.k8s.io/v1beta1","metadata":{"creationTimestamp":"2019-06-30T13:49:36Z"},"level":"Metadata","timestamp":"2019-06-30T13:49:36Z","auditID":"86a9a511-9441-493a-9d9b-a05325d30691","stage":"ResponseComplete","requestURI":"/apis/comcast.github.io/v1/namespaces/kuberhealthy/khstates/dnsstatuschecker","verb":"update","user":{"username":"system:serviceaccount:kuberhealthy:kuberhealthy","uid":"9af9ab58-99eb-11e9-b05b-02ca7bb2d18c","groups":["system:serviceaccounts","system:serviceaccounts:kuberhealthy","system:authenticated"]},"sourceIPs":["10.17.21.156"],"userAgent":"kuberhealthy/v0.0.0 (linux/amd64) kubernetes/$Format","objectRef":{"resource":"khstates","namespace":"kuberhealthy","name":"dnsstatuschecker","apiGroup":"comcast.github.io","apiVersion":"v1"},"responseStatus":{"metadata":{},"code":200},"requestReceivedTimestamp":"2019-06-30T13:49:36.172896Z","stageTimestamp":"2019-06-30T13:49:36.176632Z","annotations":{"authorization.k8s.io/decision":"allow","authorization.k8s.io/reason":"RBAC: allowed by RoleBinding \"kuberhealthy/kuberhealthy\" of Role \"kuberhealthy\" to ServiceAccount \"kuberhealthy/kuberhealthy\""}}
{"kind":"Event","apiVersion":"audit.k8s.io/v1beta1","metadata":{"creationTimestamp":"2019-06-30T13:49:36Z"},"level":"Metadata","timestamp":"2019-06-30T13:49:36Z","auditID":"d939e423-e6d9-419d-b517-20d2cc9b8888","stage":"ResponseComplete","requestURI":"/apis/comcast.github.io/v1/namespaces/kuberhealthy/khstates/componentstatuschecker","verb":"update","user":{"username":"system:serviceaccount:kuberhealthy:kuberhealthy","uid":"9af9ab58-99eb-11e9-b05b-02ca7bb2d18c","groups":["system:serviceaccounts","system:serviceaccounts:kuberhealthy","system:authenticated"]},"sourceIPs":["10.17.21.156"],"userAgent":"kuberhealthy/v0.0.0 (linux/amd64) kubernetes/$Format","objectRef":{"resource":"khstates","namespace":"kuberhealthy","name":"componentstatuschecker","apiGroup":"comcast.github.io","apiVersion":"v1"},"responseStatus":{"metadata":{},"code":200},"requestReceivedTimestamp":"2019-06-30T13:49:36.174420Z","stageTimestamp":"2019-06-30T13:49:36.177265Z","annotations":{"authorization.k8s.io/decision":"allow","authorization.k8s.io/reason":"RBAC: allowed by RoleBinding \"kuberhealthy/kuberhealthy\" of Role \"kuberhealthy\" to ServiceAccount \"kuberhealthy/kuberhealthy\""}}
{"kind":"Event","apiVersion":"audit.k8s.io/v1beta1","metadata":{"creationTimestamp":"2019-06-30T13:49:36Z"},"level":"Metadata","timestamp":"2019-06-30T13:49:36Z","auditID":"621b76ae-22eb-4420-9271-814703290359","stage":"ResponseComplete","requestURI":"/apis/comcast.github.io/v1/namespaces/kuberhealthy/khstates/podrestartchecker-namespace-kube-system","verb":"update","user":{"username":"system:serviceaccount:kuberhealthy:kuberhealthy","uid":"9af9ab58-99eb-11e9-b05b-02ca7bb2d18c","groups":["system:serviceaccounts","system:serviceaccounts:kuberhealthy","system:authenticated"]},"sourceIPs":["10.17.21.156"],"userAgent":"kuberhealthy/v0.0.0 (linux/amd64) kubernetes/$Format","objectRef":{"resource":"khstates","namespace":"kuberhealthy","name":"podrestartchecker-namespace-kube-system","apiGroup":"comcast.github.io","apiVersion":"v1"},"responseStatus":{"metadata":{},"code":200},"requestReceivedTimestamp":"2019-06-30T13:49:36.184646Z","stageTimestamp":"2019-06-30T13:49:36.188411Z","annotations":{"authorization.k8s.io/decision":"allow","authorization.k8s.io/reason":"RBAC: allowed by RoleBinding \"kuberhealthy/kuberhealthy\" of Role \"kuberhealthy\" to ServiceAccount \"kuberhealthy/kuberhealthy\""}}
{"kind":"Event","apiVersion":"audit.k8s.io/v1beta1","metadata":{"creationTimestamp":"2019-06-30T13:49:36Z"},"level":"Metadata","timestamp":"2019-06-30T13:49:36Z","auditID":"e27218fa-7a56-4bd8-88c1-6efe475fe137","stage":"ResponseComplete","requestURI":"/apis/comcast.github.io/v1/namespaces/kuberhealthy/khstates/podstatuschecker-namespace-kube-system","verb":"update","user":{"username":"system:serviceaccount:kuberhealthy:kuberhealthy","uid":"9af9ab58-99eb-11e9-b05b-02ca7bb2d18c","groups":["system:serviceaccounts","system:serviceaccounts:kuberhealthy","system:authenticated"]},"sourceIPs":["10.17.21.156"],"userAgent":"kuberhealthy/v0.0.0 (linux/amd64) kubernetes/$Format","objectRef":{"resource":"khstates","namespace":"kuberhealthy","name":"podstatuschecker-namespace-kube-system","apiGroup":"comcast.github.io","apiVersion":"v1"},"responseStatus":{"metadata":{},"code":200},"requestReceivedTimestamp":"2019-06-30T13:49:36.189272Z","stageTimestamp":"2019-06-30T13:49:36.192014Z","annotations":{"authorization.k8s.io/decision":"allow","authorization.k8s.io/reason":"RBAC: allowed by RoleBinding \"kuberhealthy/kuberhealthy\" of Role \"kuberhealthy\" to ServiceAccount \"kuberhealthy/kuberhealthy\""}}
```

If we look at the `get` requests on the `5d6` API, we can see see that the previous iteration had resulted in 2 authorized `get` reqs and the 2nd only one... So I guess it crashed around about there between where the two would have happened?  Or perhaps we're getting a prom scrape at the same time?

```json
 $ cloudwatch_log_grabber.sh  "shuttr.*cluster" "kube-apiserver-audit-5d6.*" "2019-06-30T13:49:20Z" "2019-06-30T13:49:45Z" | sed 1,2d | jq -r '.events[].message | fromjson | select(.objectRef.apiGroup == "comcast.github.io") | select(.verb == "get")' | jq -c 
{"kind":"Event","apiVersion":"audit.k8s.io/v1beta1","metadata":{"creationTimestamp":"2019-06-30T13:49:20Z"},"level":"Metadata","timestamp":"2019-06-30T13:49:20Z","auditID":"b6d3574a-751c-46b1-b9b3-d83e94324d78","stage":"ResponseComplete","requestURI":"/apis/comcast.github.io/v1/namespaces/kuberhealthy/khstates/dnsstatuschecker","verb":"get","user":{"username":"system:serviceaccount:kuberhealthy:kuberhealthy","uid":"9af9ab58-99eb-11e9-b05b-02ca7bb2d18c","groups":["system:serviceaccounts","system:serviceaccounts:kuberhealthy","system:authenticated"]},"sourceIPs":["10.17.21.156"],"userAgent":"kuberhealthy/v0.0.0 (linux/amd64) kubernetes/$Format","objectRef":{"resource":"khstates","namespace":"kuberhealthy","name":"dnsstatuschecker","apiGroup":"comcast.github.io","apiVersion":"v1"},"responseStatus":{"metadata":{},"code":200},"requestReceivedTimestamp":"2019-06-30T13:49:20.603201Z","stageTimestamp":"2019-06-30T13:49:20.605696Z","annotations":{"authorization.k8s.io/decision":"allow","authorization.k8s.io/reason":"RBAC: allowed by RoleBinding \"kuberhealthy/kuberhealthy\" of Role \"kuberhealthy\" to ServiceAccount \"kuberhealthy/kuberhealthy\""}}
{"kind":"Event","apiVersion":"audit.k8s.io/v1beta1","metadata":{"creationTimestamp":"2019-06-30T13:49:20Z"},"level":"Metadata","timestamp":"2019-06-30T13:49:20Z","auditID":"de0c95aa-1938-4147-9c8d-b4c852cc8c16","stage":"ResponseComplete","requestURI":"/apis/comcast.github.io/v1/namespaces/kuberhealthy/khstates/dnsstatuschecker","verb":"get","user":{"username":"system:serviceaccount:kuberhealthy:kuberhealthy","uid":"9af9ab58-99eb-11e9-b05b-02ca7bb2d18c","groups":["system:serviceaccounts","system:serviceaccounts:kuberhealthy","system:authenticated"]},"sourceIPs":["10.17.21.156"],"userAgent":"kuberhealthy/v0.0.0 (linux/amd64) kubernetes/$Format","objectRef":{"resource":"khstates","namespace":"kuberhealthy","name":"dnsstatuschecker","apiGroup":"comcast.github.io","apiVersion":"v1"},"responseStatus":{"metadata":{},"code":200},"requestReceivedTimestamp":"2019-06-30T13:49:20.606204Z","stageTimestamp":"2019-06-30T13:49:20.608532Z","annotations":{"authorization.k8s.io/decision":"allow","authorization.k8s.io/reason":"RBAC: allowed by RoleBinding \"kuberhealthy/kuberhealthy\" of Role \"kuberhealthy\" to ServiceAccount \"kuberhealthy/kuberhealthy\""}}
{"kind":"Event","apiVersion":"audit.k8s.io/v1beta1","metadata":{"creationTimestamp":"2019-06-30T13:49:35Z"},"level":"Metadata","timestamp":"2019-06-30T13:49:35Z","auditID":"439cbb9d-4661-47dc-8ae2-eafdca545b13","stage":"ResponseComplete","requestURI":"/apis/comcast.github.io/v1/namespaces/kuberhealthy/khstates/dnsstatuschecker","verb":"get","user":{"username":"system:serviceaccount:kuberhealthy:kuberhealthy","uid":"9af9ab58-99eb-11e9-b05b-02ca7bb2d18c","groups":["system:serviceaccounts","system:serviceaccounts:kuberhealthy","system:authenticated"]},"sourceIPs":["10.17.21.156"],"userAgent":"kuberhealthy/v0.0.0 (linux/amd64) kubernetes/$Format","objectRef":{"resource":"khstates","namespace":"kuberhealthy","name":"dnsstatuschecker","apiGroup":"comcast.github.io","apiVersion":"v1"},"responseStatus":{"metadata":{},"code":200},"requestReceivedTimestamp":"2019-06-30T13:49:35.605159Z","stageTimestamp":"2019-06-30T13:49:35.607543Z","annotations":{"authorization.k8s.io/decision":"allow","authorization.k8s.io/reason":"RBAC: allowed by RoleBinding \"kuberhealthy/kuberhealthy\" of Role \"kuberhealthy\" to ServiceAccount \"kuberhealthy/kuberhealthy\""}}
```

**Comment by integrii** [comment]
Based on some results from @rshutt, it looks like the recently patch (with tons of mutexting) to the CRD package helped, but we can still have crashes.

I think updating to the newest version of `apimachinery` could fix this.

**Comment by integrii** [comment]
I think I have a fix for this (again).  I moved the system to use a single client, rather than creating clients all over the place.  I believe something  inside apimachinery does not fully support running multiple clients concurrently.  The fix is in the 2.0.0 branch.

I am leaving this open for now, but in testing this is looking very fixed.

**Comment by integrii** [comment]
This is showing as fixed in version 2 in our testing.  No crashes at all with lots of CRD use.
